container.go

```diff
@@ -481,41 +481,40 @@ func (c *Container) stop() error {
 	// In case our container is "ready", there is no point in trying to
 	// stop it because nothing has been started. However, this is a valid
 	// case and we handle this by updating the container state only.
-	if state.State == StateReady {
-		if err := state.validTransition(StateReady, StateStopped); err != nil {
-			return err
-		}
-
-		return c.setContainerState(StateStopped)
-	}
 
-	if state.State != StateRunning {
+	if state.State != StateRunning && state.State != StateReady {
 		return fmt.Errorf("Container not running, impossible to stop")
 	}
 
-	err = state.validTransition(StateRunning, StateStopped)
+	err = state.validTransition(state.State, StateStopped)
 	if err != nil {
 		return err
 	}
 
-	if _, _, err := c.pod.proxy.connect(*(c.pod), false); err != nil {
-		return err
-	}
-	defer c.pod.proxy.disconnect()
+	if state.State == StateRunning {
+		if _, _, err := c.pod.proxy.connect(*(c.pod), false); err != nil {
+			return err
+		}
+		defer c.pod.proxy.disconnect()
 
-	err = c.pod.agent.killContainer(*(c.pod), *c, syscall.SIGKILL, true)
-	if err != nil {
-		return err
+		if err := c.pod.agent.killContainer(*(c.pod), *c, syscall.SIGKILL, true); err != nil {
+			return err
+		}
+	} else if state.State == StateReady {
+		if err := stopShim(c.process.Pid); err != nil {
```

Comment:
Can you add some comments regarding the ready case explaining the flow from shim to proxy, and proxy sending the return code here. Will be easier to follow in the future.